raft_node.py

```diff
@@ -0,0 +1,595 @@
+                    if (
+                        replications
+                        > len(self.config.all_raft_server_ports) / 2
+                    ):
+                        return
+
+    def _update_commit_index(self):
+        """We can update our commit index once there is a higher index
+        that's been replicated on a majority of servers.
+        """
+        assert isinstance(self._state, LeaderState)
+        match_indices = sorted(self._state.match_index.values())
+
+        # This implicitly counts our own commit index as the highest value
+        halfway_point = int(len(self.config.all_raft_server_ports) / 2)
+
+        min_with_majority = match_indices[halfway_point]
+        if (
+            min_with_majority > self._log.commit_index()
+            and self._log.term_for_index(min_with_majority)
+            == self._current_term
+        ):
+            self._print(f"New commit index: {min_with_majority}")
+            self._log.set_commit_index(min_with_majority)
+
+    async def _propose_mapping_handler(
+        self: Self, mapping: Mapping
+    ) -> ProposeMappingResponse:
+        """If we aren't the leader, let the proposer know who they should
+        talk to. Otherwise append the new mapping locally and attempt to
+        apply it to the state machine by sending AppendEntries requests to
+        all other nodes.
+        """
+        match self._state:
+            case FollowerState(leader=leader):
+                return ProposeMappingResponse(
+                    success=False, current_leader=leader
+                )
+            case CandidateState():
+                return ProposeMappingResponse(
+                    success=False, current_leader=None
+                )
+            case LeaderState():
+                self._print(
+                    "Received mapping proposal: "
+                    f"{mapping.key}={mapping.value}"
+                )
+                self._log.add_new_entry(
+                    term=self._current_term,
+                    key=mapping.key,
+                    value=mapping.value,
+                )
+                await self._append_entries()
+                self._print(
+                    f"Mapping {mapping.key}={mapping.value} "
+                    "successfully replicated; updating commit index"
+                )
+                self._update_commit_index()
+                return ProposeMappingResponse(
+                    success=True, current_leader=self.config.raft_server_port
+                )
+            case _:
+                assert False
+
+    def _become_follower(self: Self, new_leader: int | None) -> None:
+        """A follower will remain as such for as long as a leader is
+        heartbeating. If the heartbeat loop ever times out, become
+        a candidate.
+        """
+        match self._state:
+            case FollowerState(leader) if leader is not None:
+                self._state.leader = new_leader
```

Comment:
This case seems weird to me - it feels like it preserves your heartbeat from the old leader, or something like that, but that just seems worse than tearing down and creating a new state